matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/DefaultCryptoService.kt

```diff
@@ -907,27 +649,197 @@ internal class DefaultCryptoService @Inject constructor(
         }
     }
 
-    /**
-     * Upload my user's device keys.
-     */
-    private suspend fun uploadDeviceKeys() {
-        if (cryptoStore.areDeviceKeysUploaded()) {
-            Timber.d("Keys already uploaded, nothing to do")
-            return
+    private fun notifyRoomKeyReceived(
+            roomId: String,
+            sessionId: String,
+    ) {
+        // The sender key is actually unused since it's unimportant for megolm
+        // Our events don't contain the info so pass an empty string until we
+        // change the listener definition
+        val senderKey = ""
+
+        newSessionListeners.forEach {
+            try {
+                it.onNewSession(roomId, senderKey, sessionId)
+            } catch (e: Throwable) {
+            }
         }
-        // Prepare the device keys data to send
-        // Sign it
-        val canonicalJson = JsonCanonicalizer.getCanonicalJson(Map::class.java, getMyDevice().signalableJSONDictionary())
-        var rest = getMyDevice().toRest()
+    }
+
+    suspend fun receiveSyncChanges(
+            toDevice: ToDeviceSyncResponse?,
+            deviceChanges: DeviceListResponse?,
+            keyCounts: DeviceOneTimeKeysCountSyncResponse?) {
+        // Decrypt and handle our to-device events
+        val toDeviceEvents = this.olmMachine.receiveSyncChanges(toDevice, deviceChanges, keyCounts)
+
+        // Notify the our listeners about room keys so decryption is retried.
+        if (toDeviceEvents.events != null) {
```

Comment:
Would it be possible to have an API from olmMachine that would fire some events like `ROOM_KEY_IMPORTED`and `SECRET_RECEIVED`.

It woud be particularly usefull for secrets, as the rust-sdk could then give also the secret name. Because actually we only have the secret value, so no idea what to do with it (here for example the ssk/usk/msk seems to be sent to the keybackup service)

Reply:
Firing events, while not impossible, would:
1. Require us to pass in callbacks over FFI
2. Mess with the order of execution, i.e. the callback would probably fire before we have everything handled on the Rust side
3. Introduce dissonance into the simple push/pull design

The `receiveSyncChanges` method, of course, can return arbitrary data, so we can include extra info you need in the return value.

What's exactly needed for room keys? I agree that the secret send thing sucks and may need additional data.

Reply:
> What's exactly needed for room keys?

It's used by the opened timeline. When a new megolm session is received it will trigger a refresh (will retry to decrypt UISIs) to display properly the event on screen.

Reply:
> It's used by the opened timeline. When a new megolm session is received it will trigger a refresh (will retry to decrypt UISIs) to display properly the event on screen.

I know that, but isn't that already happening? The return value of `receiveSyncChanges` contains the decrypted to-device events. `m.room_key` events will be there as well, just that the `session_key` will be an empty string.

Am I missing something? Isn't the notifying happening right here: https://github.com/poljar/element-android/blob/rust-verification/matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/DefaultCryptoService.kt#L758.

Reply:
Yes it's working fine for new session in this case. But I am not sure it will work when i am getting a key from backup in background. Would have to also call notifyRoomKeyReceived after a keybackup import

Reply:
Ah OK. 

I think it makes more sense to modify the return value of `OlmMachine.importKeys` and `OlmMachine.importDecryptedKeys` to tell us which room keys were imported than to add callbacks that get called when the Rust side receives a room key.

Would that work?

Reply:
Yes would allow me to notify from cryptoService. I would have to refactor a bit cause now only the DefaultCryptoService is aware of the listeners, the KeybackupService is not and can't reference back the cryptoservice. Could introduced a new class of session scope to handle that and be injected in both

Reply:
Importing now returns a map of keys that were imported, see this commit: https://github.com/vector-im/element-android/pull/3209/commits/d138306b0826fb81ae5922c9c595012760ba6bfd.

You just need to traverse the map after the import and call the notifiers.